plugin/evm/mempool.go

```diff
@@ -0,0 +1,210 @@
+		return nil
+	}
+
+	m.txs[txID] = tx
+	return nil
+}
+
+// NextTx returns a transaction to be issued from the mempool.
+func (m *Mempool) NextTx() (*Tx, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for txID, tx := range m.txs {
+		delete(m.txs, txID)
+		m.currentTx = tx
+		// If there are still atomic transactions remaining
+		// ensure there is an item on the pending channel.
+		if len(m.txs) > 0 {
+			m.addPending()
+		}
+		return tx, true
+	}
+
+	return nil, false
+}
+
+// GetTx returns the transaction [txID] if it was issued
+// by this node and returns whether it was dropped and whether
+// it exists.
+func (m *Mempool) GetTx(txID ids.ID) (*Tx, bool, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if tx, ok := m.txs[txID]; ok {
+		return tx, false, true
+	}
+	if tx, ok := m.issuedTxs[txID]; ok {
+		return tx, false, true
+	}
+	if m.currentTx != nil && m.currentTx.ID() == txID {
+		return m.currentTx, false, true
+	}
+	if tx, exists := m.discardedTxs.Get(txID); exists {
+		return tx.(*Tx), true, true
+	}
+
+	return nil, false, false
+}
+
+// IssueCurrentTx marks [currentTx] as issued if there is one
+func (m *Mempool) IssueCurrentTx() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.currentTx != nil {
+		m.issuedTxs[m.currentTx.ID()] = m.currentTx
+		m.currentTx = nil
+	}
+
+	if len(m.txs) > 0 {
+		m.addPending()
+	}
+}
+
+// CancelCurrentTx marks the attempt to issue [currentTx]
+// as being aborted. If this is called after a buildBlock error
+// caused by the atomic transaction, then DiscardCurrentTx should have been called
+// such that this call will have no effect and should not re-issue the invalid tx.
+func (m *Mempool) CancelCurrentTx() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.currentTx != nil {
+		m.txs[m.currentTx.ID()] = m.currentTx
+		m.currentTx = nil
+	}
+
+	// If the VM failed to build a block for any reason
+	// make sure there is an item on the Pending channel if
+	// there are any more transactions to issue.
+	if len(m.txs) > 0 {
+		m.addPending()
+	}
+}
+
+// DiscardCurrentTx marks [currentTx] as invalid and aborts the attempt
+// to issue it since it failed verification.
+func (m *Mempool) DiscardCurrentTx() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.currentTx == nil {
+		return
+	}
+
+	m.discardedTxs.Put(m.currentTx.ID(), m.currentTx)
+	m.currentTx = nil
+}
+
+// RemoveTx removes [txID] from the mempool completely.
+func (m *Mempool) RemoveTx(txID ids.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.currentTx != nil && m.currentTx.ID() == txID {
+		m.currentTx = nil
+	}
+	delete(m.txs, txID)
+	delete(m.issuedTxs, txID)
+}
+
+// RejectTx marks [txID] as being rejected and attempts to re-issue
+// it if it was previously in the mempool.
+func (m *Mempool) RejectTx(txID ids.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	tx, ok := m.issuedTxs[txID]
+	if !ok {
+		return
+	}
+	// If the transaction was issued by the mempool, add it back
+	// to transactions pending issuance.
```

Comment:
Would we also want to remove from `issuedTxs` here if we are adding back into the mempool?